periodic/periodic_test.go

```diff
@@ -152,3 +171,88 @@ func TestID(t *testing.T) {
 		}
 	}
 }
+
+func TestInfiniteDurationAndAbort(t *testing.T) {
+	var count int64
+	var lock sync.Mutex
+	c := TestCount{&count, &lock}
+	o := RunnerOptions{
+		QPS:        10,
+		NumThreads: 1,
+		Duration:   -1, // infinite but we'll abort after 1sec
+	}
+	r := NewPeriodicRunner(&o)
+	r.Options().MakeRunners(&c)
+	count = 0
+	go func() {
+		time.Sleep(1 * time.Second)
+		log.LogVf("Calling abort after 1 sec")
+		r.Options().Abort()
+	}()
+	r.Run()
+	if count < 9 || count > 12 {
+		t.Errorf("Test executed unexpected number of times %d instead of 9-12", count)
+	}
+	// Same with infinite qps
+	count = 0
+	o.QPS = -1 // infinite qps
+	r = NewPeriodicRunner(&o)
+	r.Options().MakeRunners(&c)
+	go func() {
+		time.Sleep(140 * time.Millisecond)
+		log.LogVf("Sending global interrupt after 0.14 sec")
+		gAbortChan <- os.Interrupt
+	}()
+	r.Run()
+	if count != 3 { // should get 3 in 140ms
+		t.Errorf("Test executed unexpected number of times %d instead of %d", count, 3)
+	}
+}
+
+func TestSleepFallingBehind(t *testing.T) {
+	var count int64
+	var lock sync.Mutex
+	c := TestCount{&count, &lock}
+	o := RunnerOptions{
+		QPS:        1000000, // similar to max qps but with sleep falling behind
+		NumThreads: 4,
+		Duration:   140 * time.Millisecond,
+	}
+	r := NewPeriodicRunner(&o)
+	r.Options().MakeRunners(&c)
+	count = 0
+	res := r.Run()
+	expected := int64(3 * 4) // can start 3 50ms in 140ms * 4 threads
+	// Check the count both from the histogram and from our own test counter:
+	actual := res.DurationHistogram.Count
+	if actual != expected {
+		t.Errorf("Extra high qps executed unexpected number of times %d instead %d", actual, expected)
+	}
+	if count != expected {
+		t.Errorf("Extra high qps executed unexpected number of times %d instead %d", count, expected)
+	}
+}
+
+func Test2Watchers(t *testing.T) {
+	time.Sleep(200 * time.Millisecond)
+	{
+		o1 := RunnerOptions{}
+		r1 := newPeriodicRunner(&o1)
+		o2 := RunnerOptions{}
+		r2 := newPeriodicRunner(&o2)
+		time.Sleep(200 * time.Millisecond)
+		gAbortMutex.Lock()
+		if gOutstandingRuns != 2 {
+			t.Errorf("expecting 2 watches, found %d for (%v %v)", gOutstandingRuns, r1, r2)
```

Comment:
super nit: typical go style is `got, want` (as opposed to `want, got`)

Reply:
fixing